notarization_poller/src/notarization_poller/task.py

```diff
@@ -0,0 +1,321 @@
+#!/usr/bin/env python
+"""Task execution.
+
+Attributes:
+    NOTARIZATION_POLL_REGEX: the regex to find the notarization status
+    Task (object): the task object
+    log (logging.Logger): the log object for the module
+
+"""
+
+import asyncio
+import gzip
+import logging
+import os
+import pprint
+import re
+import signal
+import traceback
+
+import aiohttp
+import arrow
+import async_timeout
+import taskcluster
+import taskcluster.exceptions
+from taskcluster.aio import Queue
+from taskcluster.client import createTemporaryCredentials
+
+from notarization_poller.constants import get_reversed_statuses
+from notarization_poller.exceptions import RetryError
+from scriptworker_client.aio import download_file, retry_async
+from scriptworker_client.constants import STATUSES
+from scriptworker_client.exceptions import Download404, DownloadError, TaskError
+from scriptworker_client.utils import load_json_or_yaml, makedirs, rm, run_command
+
+log = logging.getLogger(__name__)
+NOTARIZATION_POLL_REGEX = re.compile(r"Status: (?P<status>success|invalid)")
+
+
+class Task:
+    """Manages all information related to a single running task."""
+
+    reclaim_fut = None
+    task_fut = None
+    complete = False
+    process = None
+    uuids = None
+
+    def __init__(self, config, claim_task, event_loop=None):
+        """Initialize Task."""
+        self.config = config
+        self.task_id = claim_task["status"]["taskId"]
+        self.run_id = claim_task["runId"]
+        self.claim_task = claim_task
+        self.event_loop = event_loop or asyncio.get_event_loop()
+        self.task_dir = os.path.join(config["work_dir"], self.task_id)
+        self.log_path = os.path.join(self.task_dir, "live_backing.log")
+        self.poll_log_path = os.path.join(self.task_dir, "polling.log")
+        rm(self.task_dir)
+        makedirs(self.task_dir)
+        self.reclaim_fut = event_loop.create_task(self.reclaim_task())
+        self._reclaim_task = {}
+        self.task_fut = event_loop.create_task(self.run_task())
+
+    @property
+    def task_credentials(self):
+        """Return the temporary credentials returned from [re]claimWork."""
+        return self._reclaim_task.get("credentials", self.claim_task["credentials"])
+
+    async def reclaim_task(self):
+        """Try to reclaim a task from the queue.
+
+        This is a keepalive.  Without it the task will expire and be re-queued.
+
+        A 409 status means the task has been resolved.
+
+        Raises:
+            TaskclusterRestFailure: on non-409 status_code from
+                `taskcluster.aio.Queue.reclaimTask`
+
+        """
+        while True:
+            log.debug("waiting %s seconds before reclaiming..." % self.config["reclaim_interval"])
+            await asyncio.sleep(self.config["reclaim_interval"])
```

Comment:
No, we should have a config option that is significantly less than the `takenUntil`, or a slightly delayed `reclaimTask` will result in a `claim-expired` task that gets rerun. This is the same model as scriptworker.

Reply:
I agree that the time used should be less than `takenUntil`, not equal to it. But, since we ignore the value that taskcluster returns, we could in fact have an interval that is past `takenUntil` time. Presumably this would require a change to the taskcluster configuration, but there isn't any reason we can't be resilient to changes to that. If we do have a config, it should be for how long before `takenUntil` we should reclaim, rather than how long between reclaims, or something along those lines.

Reply:
The current config-driven reclaim interval mirrors what we do in scriptworker, which is what the taskcluster team told me to do when I first wrote it. I'm inclined to leave that in place for now.

Reply:
Looking at the taskcluster maintained workers:
- docker-worker [reclaims](https://github.com/taskcluster/docker-worker/blob/b089a1dbf5aed5b9cd1f70fcf25f1e92ff05a15a/src/lib/task.js#L183) at 3/4 of the time between now and `takenUntil` (`reclaimDivisor` is 1.3 and 1/1.3 is approximately 76%)
- generic-worker [reclaims](https://github.com/taskcluster/generic-worker/blob/5fd8e48d63d2f7a7eeb6c9a65d4c4d1b41d535b7/taskstatus.go#L340) 3m before `takenUntil`, with some protection from hitting the queue too often

I think we should do something similar here (and eventually update scriptworker). I don't know if the advice changed, or maybe taskcluster didn't used to report claim times and so picking an interval made sense when scriptworker was originally written.

Reply:
Tracking followup in https://github.com/mozilla-releng/scriptworker/issues/424 .